processor/transformprocessor/internal/metrics/func_convert_exponential_hist_to_explicit_hist.go

```diff
@@ -0,0 +1,305 @@
+		return nil, fmt.Errorf("explicit bounds cannot be empty: %v", explicitBounds)
+	}
+
+	distFn, ok := distributionFnMap[distributionFn]
+	if !ok {
+		return nil, fmt.Errorf("invalid distribution alogrithm: %s, must be one of [upper, midpoint, random, uniform]", distributionFn)
+	}
+
+	return func(_ context.Context, tCtx ottlmetric.TransformContext) (any, error) {
+		metric := tCtx.GetMetric()
+
+		// only execute on exponential histograms
+		if metric.Type() != pmetric.MetricTypeExponentialHistogram {
+			return nil, nil
+		}
+
+		explicitHist := pmetric.NewHistogram()
+		dps := metric.ExponentialHistogram().DataPoints()
+		explicitHist.SetAggregationTemporality(metric.ExponentialHistogram().AggregationTemporality())
+
+		// map over each exponential histogram data point and calculate the bucket counts
+		for i := 0; i < dps.Len(); i++ {
+			expDataPoint := dps.At(i)
+			bucketCounts := calculateBucketCounts(expDataPoint, explicitBounds, distFn)
+			explicitHistDp := explicitHist.DataPoints().AppendEmpty()
+			explicitHistDp.SetStartTimestamp(expDataPoint.StartTimestamp())
+			explicitHistDp.SetTimestamp(expDataPoint.Timestamp())
+			explicitHistDp.SetCount(expDataPoint.Count())
+			explicitHistDp.SetSum(expDataPoint.Sum())
+			explicitHistDp.SetMin(expDataPoint.Min())
+			explicitHistDp.SetMax(expDataPoint.Max())
+			expDataPoint.Exemplars().CopyTo(explicitHistDp.Exemplars())
+			explicitHistDp.ExplicitBounds().FromRaw(explicitBounds)
+			explicitHistDp.BucketCounts().FromRaw(bucketCounts)
+			expDataPoint.Attributes().CopyTo(explicitHistDp.Attributes())
+		}
+
+		// create new metric and override metric
+		newMetric := pmetric.NewMetric()
+		newMetric.SetName(metric.Name())
+		newMetric.SetDescription(metric.Description())
+		newMetric.SetUnit(metric.Unit())
+		explicitHist.CopyTo(newMetric.SetEmptyHistogram())
+		newMetric.CopyTo(metric)
+
+		return nil, nil
+	}, nil
+}
+
+type distAlgorithm func(count uint64, upper, lower float64, boundaries []float64, bucketCountsSrc pcommon.UInt64Slice, bucketCountsDst *[]uint64)
+
+func calculateBucketCounts(dp pmetric.ExponentialHistogramDataPoint, boundaries []float64, distFn distAlgorithm) []uint64 {
+	scale := int(dp.Scale())
+	factor := math.Ldexp(math.Ln2, -scale)
+	posB := dp.Positive().BucketCounts()
+	bucketCounts := make([]uint64, len(boundaries))
+
+	// add zerocount if boundary starts at zero
+	if zerocount := dp.ZeroCount(); zerocount > 0 && boundaries[0] == 0 {
+		bucketCounts[0] += zerocount
+	}
+
+positionLoop:
+	for pos := 0; pos < posB.Len(); pos++ {
+		index := dp.Positive().Offset() + int32(pos)
+		upper := math.Exp(float64(index+1) * factor)
+		lower := math.Exp(float64(index) * factor)
+		count := posB.At(pos)
+
+		// check if lower and upper bounds are within the boundaries
+		for bIndex := 1; bIndex < len(boundaries); bIndex++ {
+			if lower > boundaries[bIndex-1] && upper <= boundaries[bIndex] {
+				bucketCounts[bIndex-1] += count
+				continue positionLoop
+			}
+		}
```

Comment:
We have a design principle (technically for standard OTTL functions, which this is not, but I still like to point to them) that requires that OTTL functions be [terminating](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs/README.md#design-principles).  Is there a way to refactor this loop to not need a GOTO? The value instead the if statement isn't used in the condition, so is it possible to get stuck continuously being sent back to the start of the first for loop?

Reply:
Wrapped the `distFn` call in a condition, changed the `continue` to `break` and removed the GOTO